### ISSUE TITLE
Infer number of labels in RecordReaderDataSetIterator from recordreader

### ISSUE THREAD

**Issue by agibsonccc** [deeplearning4j/deeplearning4j#3216]
#### Issue Description
Right now we pass in numLabels for the constructor:
https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j-core/src/main/java/org/deeplearning4j/datasets/datavec/RecordReaderDataSetIterator.java#L256

Please describe our issue, along with:
- expected behavior
Infer labels from record reader
- encountered behavior
Having to pass in the constructor
#### Version Information
master
Please indicate relevant versions, including, if relevant:

* Deeplearning4j version
* platform information (OS, etc)
* CUDA version, if used
* NVIDIA driver version, if in use

#### Contributing

If you'd like to help us fix the issue by contributing some code, but would
like guidance or help in doing so, please mention it!

**Comment by AlexDBlack** [comment]
Sounds good. We'd need to be careful here though:
- Some things like CSVs: might (a) need a full dataset pass to infer this, and (b) in some cases (like k-fold validation, test sets etc) we might not have every class present in the training data
- For ImageRecordReader (and perhaps some others), it should work well
- Overall: log that ```N labels were inferred from RecordReader, use X to override``` or similar

**Comment by raver119** [comment]
Is this issue still viable?
